### ISSUE TITLE
ConsensusID tool writes incorrect float values

### ISSUE THREAD

**Issue by lars20070** [OpenMS/OpenMS#2020]
The `ConsensusID` tool writes values such as `value="1.#INF"`. See for example.
```
<PeptideIdentification score_type="XTandem" higher_score_better="true" significance_threshold="0" MZ="484.821270750295" RT="1573.41798" >
	<PeptideHit score="16.3" sequence="(Dimethyl:2H(6)13C(2))YLSPLALR" charge="2" >
		<UserParam type="float" name="consensus_support" value="1.#INF"/>
	</PeptideHit>
</PeptideIdentification>
```
which leads to a crash/exception in the following `PeptideIndexer`.
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/DATASTRUCTURES/StringUtils.h#L727

The workflow did run under Windows.

**Comment by lars20070** [comment]
The bug occurs only under Windows.

The `ConsensusID` tool writes `inf` and `nan` float values. But under Windows the `inf` values are incorrectly written as `1.#INF` which leads to a crash in subsequent tools [1].

@all Do we really want to write NaNs and infinities? Or should we ensure that the tools write only finite floats?

[1] http://www.johndcook.com/blog/ieee_exceptions_in_cpp/

**Comment by hendrikweisser** [comment]
Do the "inf" and "nan" values occur only in the "consensus_support" meta data? If yes, then I think they must be due to a bug. The support value is supposed to tell you what fraction of other search engines supported an ID, so it should really be between 0 and 1.

**Comment by hendrikweisser** [comment]
What settings did you use in ConsensusID when you got those values?

**Comment by lars20070** [comment]
Thought so. Just the `best` algorithm. Three separate searches from the same search engine are combined.

**Comment by lars20070** [comment]
`ConsensusID (best)` on these three MS-GF+ searches [1].

[1] https://gist.github.com/lars20070/71c9c264ecaa903b1ca7026c77dbf27a

**Comment by hendrikweisser** [comment]
Can you check if the error still occurs if you set the "filter:count_empty" flag of ConsensusID?

**Comment by lars20070** [comment]
Nope. With the `count_empty` flag all `consensus_support` are 0.

**Comment by lars20070** [comment]
Bug occurs with XTandem but not MS-GF+.

**Comment by hendrikweisser** [comment]
> With the count_empty flag all consensus_support are 0.

All? That would mean no two searches identified the same top hit for a spectrum. Can that be?

**Comment by lars20070** [comment]
Nearly all. - And that makes sense, since the three searches used different fixed mods, i.e. the searches are supposed to be exclusive.

**Comment by lars20070** [comment]
Thanks @hendrikweisser. The flag is a good workaround.

If the ID was seen in only one search, then the support should be zero, right?

**Comment by hendrikweisser** [comment]
> If the ID was seen in only one search, then the support should be zero, right?

That depends on your choice for "filter:count_empty". The idea behind the support value is to quantify how well multiple search runs (i.e. usually multiple different search engines) agree. If you have two search engines and they give different top hits for the same spectrum, then they clearly disagree. However, if one search engine doesn't even identify anything for the spectrum, can you say that it disagrees?

**Comment by lars20070** [comment]
In the context of `best` (identical score types from the same search engine), `min_support` should not been used at all. I cannot see a use case for `best`.

**Comment by hendrikweisser** [comment]
> In the context of best (identical score types from the same search engine), min_support should not been used at all.

Yes, I agree, but the support calculation is orthogonal to the score calculation.

**Comment by lars20070** [comment]
Guess I haven't thought this through completely. - In the context of multiple search engines, why shouldn't the support be zero if only one of the search engines reports a hit? Maybe a quick phone call tomorrow?

Thanks again. Your comments were very helpful.

**Comment by hendrikweisser** [comment]
Thanks for reporting the problem, Lars. I won't be able to do a phone call this week because I'm away at a meeting Thursday and Friday. I'll think about how to fix this bug, though.

**Comment by hendrikweisser** [comment]
By the way, I was puzzled why you would get both "inf" and "nan" values in some cases. I could understand the "nan" - it's the result of dividing "zero other runs agree" by "zero other runs are considered". But to get "inf", you would have to have a positive number of runs that agree, out of zero runs that are considered, which clearly doesn't make sense.

The reason for the "inf" is that in some cases, you get the same hit twice for a spectrum in one of your search runs (e.g. look for "(Dimethyl)TIVPIR" in your X! Tandem input file). ConsensusID doesn't expect this, sees a hit twice and thinks it must come from two search runs. I'll fix that in ConsensusID, but it might be worth checking why you get duplicated hits in your X! Tandem results.

**Comment by lars20070** [comment]
I see what you mean. As another example, the sequence `(Dimethyl)AIQGFAVGG` appears as first, second and third hit for the same MS2 spectrum. Seems the adapter or XTandem itself are still buggy.

**Comment by hendrikweisser** [comment]
I've opened a pull request that should fix the issue: #2039 
Works on your X! Tandem example file.
